### PR TITLE
Add `cache::ClearSiteData` header

### PR DESCRIPTION
**Pull request by yoshuawuyts**
Adds the `cache::ClearSiteData` header. Ref #99. Thanks!

### CODE REVIEW

File: src/cache/clear_site_data/mod.rs [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+pub use directive::ClearDirective;
+
+/// Clear browsing data (cookies, storage, cache) associated with the
+/// requesting website.
+///
+/// [MDN Documentation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Clear-Site-Data)
+///
+/// # Specifications
+///
+/// - [RFC 7232, section 3.1: If-Match](https://tools.ietf.org/html/rfc7232#section-3.1)
```
Comment (Fishrock123) [inline src/cache/clear_site_data/mod.rs:23]: This is incorrect but I also cant find an rfc for this? Maybe I don't know how to search for these rfcs...

File: src/cache/clear_site_data/mod.rs [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+        Self {
+            entries: vec![],
+            wildcard: false,
+        }
+    }
+
+    /// Create a new instance from headers.
+    pub fn from_headers(headers: impl AsRef<Headers>) -> crate::Result<Option<Self>> {
+        let mut entries = vec![];
+        let headers = match headers.as_ref().get(CLEAR_SITE_DATA) {
```
Comment (Fishrock123) [inline src/cache/clear_site_data/mod.rs:64]:
```suggestion
        let header_values = match headers.as_ref().get(CLEAR_SITE_DATA) {
```

File: src/cache/clear_site_data/mod.rs [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+                    continue;
+                }
+                entries.push(ClearDirective::from_str(part)?);
+            }
+        }
+
+        Ok(Some(Self { entries, wildcard }))
+    }
+
+    /// Sets the `If-Match` header.
```
Comment (Fishrock123) [inline src/cache/clear_site_data/mod.rs:84]:
```suggestion
    /// Sets the `Clear-Site-Data` header.
```

File: src/lib.rs [hunk 4]
```diff
@@ -94,7 +94,7 @@
 //! declare it up front. But if you don't, things will still work.
 
 #![deny(missing_debug_implementations, nonstandard_style)]
-#![warn(missing_docs, unreachable_pub)]
```
Comment (Fishrock123) [inline src/lib.rs:97]: Why this change? It should be `pub(crate)` instead then, no?

File: src/cache/clear_site_data/mod.rs (hunk 1 above)
Comment (jbr) [reply src/cache/clear_site_data/mod.rs:23]: https://w3c.github.io/webappsec-clear-site-data/ — i tend to go to mdn and then check their reference section

File: src/lib.rs (hunk 4 above)
Comment (yoshuawuyts) [reply src/lib.rs:97]: The lint triggers lots of false positives. We've gone through enabling / disabling it in the past. For whatever reason it was flagging `cache::ClearSiteData` as "private" when it's clearly public, so I think it's time to disable the lint again.

File: src/cache/clear_site_data/directive.rs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+impl ClearDirective {
+    /// Get the formatted string.
+    pub fn as_str(&self) -> &'static str {
+        match self {
+            ClearDirective::Cache => r#""cache""#,
+            ClearDirective::Cookies => r#""cookies""#,
+            ClearDirective::Storage => r#""storage""#,
+            ClearDirective::ExecutionContexts => r#""executionContexts""#,
```
Comment (jbr) [inline src/cache/clear_site_data/directive.rs:36]:
```suggestion
            ClearDirective::Cache => "cache",
            ClearDirective::Cookies => "cookies",
            ClearDirective::Storage => "storage",
            ClearDirective::ExecutionContexts => "executionContexts",
```

Why use rust string syntax when unnecessary?

File: src/cache/clear_site_data/directive.rs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+impl FromStr for ClearDirective {
+    type Err = std::string::ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            r#""cache""# => Ok(Self::Cache),
+            r#""cookies""# => Ok(Self::Cookies),
+            r#""storage""# => Ok(Self::Storage),
+            r#""executionContexts""# => Ok(Self::ExecutionContexts),
```
Comment (jbr) [inline src/cache/clear_site_data/directive.rs:55]:
```suggestion
            "cache" => Ok(Self::Cache),
            "cookies" => Ok(Self::Cookies),
            "storage" => Ok(Self::Storage),
            "executionContexts" => Ok(Self::ExecutionContexts),
```

File: src/cache/clear_site_data/mod.rs [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+        let headers = match headers.as_ref().get(CLEAR_SITE_DATA) {
+            Some(headers) => headers,
+            None => return Ok(None),
+        };
+
+        let mut wildcard = false;
+        for value in headers {
+            for part in value.as_str().trim().split(',') {
+                let part = part.trim();
+                if part == r#""*""# {
```
Comment (jbr) [inline src/cache/clear_site_data/mod.rs:73]:
```suggestion
                if part == "*'"{
```

File: src/cache/clear_site_data/directive.rs (hunk 5 above)
Comment (yoshuawuyts) [reply src/cache/clear_site_data/directive.rs:36]: The quotes here are intentional! Each value **must** include quotes, so it's strings like `"cookies", "storage"`. Hence the usage of raw strings. There's a test for this at the bottom of the file.

Comment (jbr) [reply src/cache/clear_site_data/directive.rs:36]: Ohhh! I see. Sorry!

**Comment by yoshuawuyts** [comment]
Updated with all feedback. Merging!
